# Assignment 1
### Understanding Uncertainty
### Heywood B. Williams-Tracy (heywood.williams-tracy@virginia.edu)
### Due 9/5

1. Create a new public repo on Github under your account. Include a readme file.
2. Clone it to your machine. Put this file into that repo.
3. Use the following function to download the example data for the course:

In [1]:
def download_data(force=False):
    """Download and extract course data from Zenodo."""
    import urllib.request, zipfile, os
    
    zip_path = 'data.zip'
    data_dir = 'data'
    
    if not os.path.exists(zip_path) or force:
        print("Downloading course data")
        urllib.request.urlretrieve(
            'https://zenodo.org/records/16954427/files/data.zip?download=1',
            zip_path
        )
        print("Download complete")
    else:
        print("Download file already exists")
        
    if not os.path.exists(data_dir) or force:
        print("Extracting data files...")
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            zip_ref.extractall(data_dir)
        print("Data extracted")
    else:
        print("Data directory already exists")

download_data()

Download complete
Extracting data files...
Data extracted


4. Open one of the datasets using Pandas:
    1. `ames_prices.csv`: Housing characteristics and prices
    2. `college_completion.csv`: Public, nonprofit, and for-profit educational institutions, graduation rates, and financial aid
    3. `ForeignGifts_edu.csv`: Monetary and in-kind transfers from foreign entities to U.S. educational institutions
    4. `iowa.csv`: Liquor sales in Iowa, at the transaction level
    5. `metabric.csv`: Cancer patient and outcome data
    6. `mn_police_use_of_force.csv`: Records of physical altercations between Minnessota police and private citizens
    7. `nhanes_data_17_18.csv`: National Health and Nutrition Examination Survey
    8. `tuna.csv`: Yellowfin Tuna Genome (I don't recommend this one; it's just a sequence of G, C, A, T )
    9. `va_procurement.csv`: Public spending by the state of Virginia

In [13]:
import pandas as pd

iowa_df = pd.read_csv('data/iowa.csv')
iowa_df.head()

Invoice/Item Number        Date  Store Number                   Store Name  \
0     INV-59108400026  06/06/2023          3723           J D SPIRITS LIQUOR   
1        S16879800057  01/16/2014          3926  LIQUOR DOWNTOWN / IOWA CITY   
2     INV-05301100019  06/05/2017          3829     GARY'S FOODS / MT VERNON   
3     INV-40973500083  10/14/2021          5102               WILKIE LIQUORS   
4     INV-17022500013  01/18/2019          2560   HY-VEE FOOD STORE / MARION   

  Zip Code          Category Name           Vendor Name  Item Number  \
0    51040  STRAIGHT RYE WHISKIES      INFINIUM SPIRITS        27102   
1    52240         VODKA 80 PROOF    HEAVEN HILL BRANDS        35416   
2    52314      CANADIAN WHISKIES       DIAGEO AMERICAS        11296   
3    52314      AMERICAN SCHNAPPS       JIM BEAM BRANDS        82787   
4    52302        WHISKEY LIQUEUR  SAZERAC COMPANY  INC        64863   

            Item Description  Bottle Volume (ml)  State Bottle Cost  \
0          TEMPLETON RYE 4YR                 750              18.09   
1     BURNETT'S VODKA 80 PRF                 750               4.84   
2                CROWN ROYAL                 750              15.59   
3       DEKUYPER BUTTERSHOTS                1000               7.87   
4  FIREBALL CINNAMON WHISKEY                 200               2.50   

   State Bottle Retail  Bottles Sold  Sale (Dollars)  
0                27.14             2           54.28  
1                 7.26            12           87.12  
2                23.39             6          135.66  
3                11.81            12          141.72  
4                 3.75            12           45.00

In [6]:
iowa_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159904 entries, 0 to 159903
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Invoice/Item Number  159904 non-null  object 
 1   Date                 159904 non-null  object 
 2   Store Number         159904 non-null  int64  
 3   Store Name           159904 non-null  object 
 4   Zip Code             159508 non-null  object 
 5   Category Name        159771 non-null  object 
 6   Vendor Name          159904 non-null  object 
 7   Item Number          159904 non-null  int64  
 8   Item Description     159904 non-null  object 
 9   Bottle Volume (ml)   159904 non-null  int64  
 10  State Bottle Cost    159904 non-null  float64
 11  State Bottle Retail  159904 non-null  float64
 12  Bottles Sold         159904 non-null  int64  
 13  Sale (Dollars)       159904 non-null  float64
dtypes: float64(3), int64(4), object(7)
memory usage: 17.1+ MB


5. Pick two or three variables and briefly analyze them
    - Is it a categorical or numeric variable?
    - How many missing values are there? (`df['var'].isna()` and `np.sum()`)
    - If categorical, tabulate the values (`df['var'].value_counts()`) and if numeric, get a summary (`df['var'].describe()`)

Item Description is a categorical variable. It describes the type of item sold in the transaction. Zip code is a categorical variable. It describes the location of the store where the transaction took place. Sale (Dollars) is a numeric variable. It describes the amount of money made in the transaction.

In [7]:
iowa_df['Item Description'].isna().sum()

0

In [15]:
iowa_df['Item Description'].value_counts()

BLACK VELVET                                                    3994
HAWKEYE VODKA                                                   3460
TITOS HANDMADE VODKA                                            2648
FIREBALL CINNAMON WHISKEY                                       2500
FIVE O'CLOCK VODKA                                              2013
                                                                ... 
BURNETTS BLUE RASPBERRY VODKA                                      1
MCCARTHYS 6 YEAR OREGON SINGLE MALT WHISKEY FINISHED IN PX S       1
XICARU SILVER MEZCAL                                               1
EFFEN VODKA TRIO PACK                                              1
RESILIENT BOURBON BOTTLED IN BOND                                  1
Name: Item Description, Length: 5401, dtype: int64

In [12]:
iowa_df['Zip Code'].isna().sum()

396

In [16]:
iowa_df['Zip Code'].value_counts()

50010    3831
52402    3505
52240    3324
50613    3128
50317    2836
         ... 
51533       1
52585       1
52045       1
52654       1
50050       1
Name: Zip Code, Length: 517, dtype: int64

In [14]:
iowa_df['Sale (Dollars)'].isna().sum()

0

In [17]:
iowa_df['Sale (Dollars)'].describe()

count    159904.000000
mean        146.931598
std         459.368723
min       -1628.400000
25%          36.550000
50%          78.660000
75%         152.880000
max       36720.000000
Name: Sale (Dollars), dtype: float64

There are no missing values in either the item description or the sale amount but there are 396 missing entries in zip code.

6. What are some questions and prediction tools you could create using these data? Who would the stakeholder be for that prediction tool? What practical or ethical questions would it create? What other data would you want, that are not available in your data?

Alcohol sales trend over time depending on the time of year, weather and location. By being able to predict which products will sell best where and when stories can optimize their inventory so no customer has to leave empty-handed. The Iowa state ABC board would be a stakeholder in this prediction tool. Practical questions would include how to best stock stores to meet customer demand while minimizing waste. Ethical questions could include ensuring that the data used to make predictions does not inadvertently promote overconsumption of alcohol or target vulnerable populations. Additional data that would be useful include weather, local event schedules, and demographic data about the area around where the stores are located.

7. Commit your work to the repo (`git commit -am 'Finish assignment'` at the command line, or use the Git panel in VS Code). Push your work back to Github and submit the link on Canvas in the assignment tab.